src/sentry/unmerge.py

```diff
@@ -0,0 +1,226 @@
+
+    @abc.abstractmethod
+    def get_unmerge_key(
+        self, event: Event, locked_primary_hashes: Collection[str]
+    ) -> Optional[str]:
+        """
+        The unmerge task iterates through all events of a group. This function
+        should return which of them should land in the new group.
+
+        If the issue should be moved, a string should be returned. Events with
+        the same string are moved into the same issue.
+        """
+
+        raise NotImplementedError()
+
+    @abc.abstractproperty
+    def primary_hashes_to_lock(self) -> Collection[str]:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def start_snuba_replacement(self, project: Project, source_id: int, destination_id: int) -> Any:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def stop_snuba_replacement(self, eventstream_state: Any) -> None:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def run_postgres_replacement(
+        self, project: Project, destination_id: int, locked_primary_hashes: Collection[str]
+    ) -> None:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_activity_args(self) -> Mapping[str, Any]:
+        raise NotImplementedError()
+
+
+@dataclass(frozen=True)
+class PrimaryHashUnmergeReplacement(UnmergeReplacement):
+    """
+    The "classical unmerge": Moving events out of the group based on primary_hash.
+    """
+
+    fingerprints: Collection[str]
+
+    def get_unmerge_key(
+        self, event: Event, locked_primary_hashes: Collection[str]
+    ) -> Optional[str]:
+        primary_hash = event.get_primary_hash()
+        if primary_hash in self.fingerprints and primary_hash in locked_primary_hashes:
+            return _DEFAULT_UNMERGE_KEY
+
+        return None
+
+    @property
+    def primary_hashes_to_lock(self) -> Collection[str]:
+        return self.fingerprints
+
+    def start_snuba_replacement(self, project: Project, source_id: int, destination_id: int) -> Any:
+        return eventstream.start_unmerge(project.id, self.fingerprints, source_id, destination_id)
+
+    def stop_snuba_replacement(self, eventstream_state: Any) -> None:
+        eventstream.end_unmerge(eventstream_state)
+
+    def run_postgres_replacement(
+        self, project: Project, destination_id: int, locked_primary_hashes: Collection[str]
+    ) -> None:
+        # Move the group hashes to the destination.
+        GroupHash.objects.filter(project_id=project.id, hash__in=locked_primary_hashes).update(
+            group=destination_id
+        )
+
+    def get_activity_args(self) -> Mapping[str, Any]:
+        return {"fingerprints": self.fingerprints}
+
+
+_REPLACEMENT_DISCRIMINATORS: Mapping[Type[UnmergeReplacement], str] = {
+    PrimaryHashUnmergeReplacement: "primary_hash",
+}
+
+_DISCRIMINATOR_REPLACEMENTS: Mapping[str, Type[UnmergeReplacement]] = {
+    v: k for k, v in _REPLACEMENT_DISCRIMINATORS.items()
+}
```

Comment:
I guess this is because you will be adding other types in the future?

Reply:
Yup, I was trying to build my own tagged enums here.